src/plugins/cpptemplate/cpptemplate.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file
+ *
+ * @brief Source for cpptemplate plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include "cpptemplate.hpp"
+#include "cpptemplate_delegate.hpp"
+
+#include <kdbhelper.h>
+
+using elektra::CppTemplateDelegate;
+
+using CppKey = kdb::Key;
+using CppKeySet = kdb::KeySet;
+
+namespace
+{
+
+/**
+ * @brief This function returns a key set containing the plugin contract.
+ *
+ * @return A key set specifying the capabilities of the plugin
+ */
+CppKeySet getContract ()
+{
+	return CppKeySet{ 30,
+			  keyNew ("system/elektra/modules/cpptemplate", KEY_VALUE, "cpptemplate plugin waits for your orders", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/open", KEY_FUNC, elektraCppTemplateOpen, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/close", KEY_FUNC, elektraCppTemplateClose, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/get", KEY_FUNC, elektraCppTemplateGet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/set", KEY_FUNC, elektraCppTemplateSet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/error", KEY_FUNC, elektraCppTemplateError, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/checkconf", KEY_FUNC, elektraCppTemplateCheckConfig, KEY_END),
+#include ELEKTRA_README (cpptemplate)
+			  keyNew ("system/elektra/modules/cpptemplate/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END),
+			  KS_END };
+}
+
+} // end namespace
+
+extern "C" {
+
+typedef Delegator<CppTemplateDelegate> delegator;
+
+/** @see elektraDocOpen */
+int elektraCppTemplateOpen (Plugin * handle, Key * key)
+{
+	// After the call to `delegator::open` you can retrieve a pointer to the delegate via `coderDelegator::get (handle)`
+	return delegator::open (handle, key);
```

Comment:
What happens if delegator::open throws an exception? What about catching them here and set an error for uncaught exceptions?

Reply:
> What happens if delegator::open throws an exception? What about catching them here and set an error for uncaught exceptions?

The code now catches exceptions and adds warning `11`:

```
number:11
description:open of plugin returned unsuccessfully (reason contains plugin, see other warnings for details)
severity:warning
ingroup:kdb
macro:OPEN
```

in that case. I am not sure if warning `11` is really appropriate though. Should I use another warning/error instead?

Reply:
It should be an error because an error status will be returned.